Framework/Core/src/RawDeviceSource.cxx

```diff
@@ -0,0 +1,131 @@
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+#include "Framework/RawDeviceSource.h"
+#include "Framework/ProcessingContext.h"
+#include "Framework/RawDeviceService.h"
+#include "Framework/InitContext.h"
+#include "Framework/ConfigParamSpec.h"
+#include "Framework/AlgorithmSpec.h"
+#include "Headers/DataHeader.h"
+#include "Framework/DataProcessingHeader.h"
+#include <fairmq/FairMQParts.h>
+#include <cstring>
+#include <cassert>
+#include <memory>
+
+namespace o2 {
+namespace framework {
+
+using DataHeader = o2::Header::DataHeader;
+
+void broadcastMessage(FairMQDevice &device, o2::Header::Stack &&headerStack, FairMQMessagePtr &&payloadMessage, int index) {
+  for (auto &channelInfo : device.fChannels) {
+    auto channel = channelInfo.first;
+    assert(channelInfo.second.size() == 1);
+    // FIXME: I need to make sure the input channel is not used... For the moment 
+    //        I rely on the convention dpl channels start with "from_".
+    if (strncmp(channel.data(), "from_", 5) != 0) {
+      continue;
+    }
+
+    // FIXME: this assumes there is only one output from here... This should
+    //        really do the matchmaking between inputs and output channels.
+    FairMQMessagePtr headerMessage = device.NewMessageFor(channel, index,
+                                                          headerStack.buffer.get(),
+                                                          headerStack.bufferSize,
+                                                          &o2::Header::Stack::freefn,
+                                                          headerStack.buffer.get());
+    headerStack.buffer.release();
+    FairMQParts out;
+    out.AddPart(std::move(headerMessage));
+    out.AddPart(std::move(payloadMessage));
+    device.Send(out, channel, index);
+  }
+}
+
+// FIXME: should I filter out only the output specs which match?
+InjectorFunction o2DMAdaptor(OutputSpec const &spec, uint64_t startTime, uint64_t step) {
+  auto timesliceId = std::make_shared<size_t>(startTime);
+  return [timesliceId, step](FairMQDevice &device, FairMQParts &parts, int index) {
+    for (size_t i = 0; i < parts.Size()/2; ++i) {
+      auto dh = o2::Header::get<DataHeader>(parts.At(i*2)->GetData());
+
+      LOG(DEBUG) << "timesliceId: " << *timesliceId;
+      DataProcessingHeader dph{*timesliceId, 0};
+      o2::Header::Stack headerStack{*dh, dph};
+      broadcastMessage(device, std::move(headerStack), std::move(parts.At(i*2+1)), index);
+      auto oldTimesliceId = *timesliceId;
+      *timesliceId += 1;
+      assert(*timesliceId == oldTimesliceId +1);
```

Comment:
Out of curiosity why do you need this assert ? in what way could this possibly go wrong ?

Reply:
leftover from a debugging session, will remove.